subprojects/maven/src/main/java/org/gradle/api/publish/maven/tasks/GenerateMavenPom.java

```diff
@@ -118,57 +116,14 @@ public void setDestination(Object destination) {
 
     @TaskAction
     public void doGenerate() {
-        mavenPomSpec().writeTo(getDestination());
-    }
-
-    private MavenPomFileGenerator.MavenPomSpec mavenPomSpec() {
-        return mavenPomSpec.get();
+        mavenPomSpec.get().writeTo(getDestination());
```

Comment:
💭 Reading through this, this class and `MavenPomFileGenerator` separate writing the POM into generating a Spec, and then writing it.

This class uses a `Cached` spec, which is only ever accessed (and thus computed) at task execution time.

Why not just generate the Spec + write the POM all at once as a new, simpler, internal API in the `MavenPomFileGenerator` class?  It looks like all usages could just call this new single `writePom(MavenPomInternal, File)` method.  That would simplify the code a little and also remove referencing a non-static method during initialization of this class.

Reply:
I agree there is a lot of work that needs to be done in this area. I definitely plan to make more changes to this class and the POM generator in upcoming changes. 

Eventually, this class will be made configuration-cache compatible (right now it is annotated with `@UntrackedTask`), and when that happens, I don't think we should be writing files during serialization-time. I think this will still need to be split up into separate spec and file writing steps for the sake of avoiding disk IO during the configuration phase. 

The next steps in cleaning this up would be to properly mark the POM as an `@Input`.

Reply:
I'm suggesting to continue to write the files during task execution time, but to also generate the `Spec` immediately before writing the files, as an internal implementation detail of a new write method.  

I'm not sure how this would affect CC compatibility, as the file is currently written at task execution time and this wouldn't change, and the `Spec` generation doesn't do disk IO that I can see.  If anything, removing the `mavenPomSpec` field should make the `GenerateMavenPom` task more cachable.

It seemed like an quick improvement that could be done now, but if you've already got future cleanups in mind, it could certainly wait until later.

Reply:
I do believe there might be some configuration-time work going on here. If you look at `Cached`, it specifically implements `Serializable` and the `writeReplace` method. Even though this is annotated with `@UntrackedTask`, I still think CC is evaluating the `Spec` at configuration-time and writing the file at execution-time -- I saw CC errors related to this during development.

Given this, we still need to separate out the `Spec`, since the source `MavenPomInternal` is not serializable in its current form. Given the plans to update the cachability and CC compatibility of this class in future revisions, I will not make additional changes in this PR.